### ISSUE TITLE
ApiGatewayWebSocket custom authorizer

### ISSUE THREAD

**Issue by wilfredjonathanjames** [anomalyco/sst#5089]
Plans for custom authorizer on `ApiGatewayWebSocket`? Only supports `iam` currently.

Looking into work-arounds. Will update if I find any.

Thanks again for sst - has sped up infra and dev time.

**Comment by wilfredjonathanjames** [comment]
Docs [here](https://sst.dev/docs/component/aws/apigateway-websocket#auth).

**Comment by wilfredjonathanjames** [comment]
This is as far as I've got, but I'm blocked by the following error

```typescript
/// <reference path="../.sst/platform/config.d.ts" />

import { ApiGatewayV2Authorizer } from "../.sst/platform/src/components/aws/apigatewayv2-authorizer"

export default async function configureWebSockets(
  vpc: sst.aws.Vpc,
  db: sst.aws.Postgres,
) {
  const api = new sst.aws.ApiGatewayWebSocket("WebSocketApi", {
    transform: {
      route: {
        handler: {
          link: [db],
          vpc,
        },
      },
    },
  })

  const authorizer = new ApiGatewayV2Authorizer("WebSocketApiAuthorizer", {
    name: "WebSocketApiAuthorizer",
    api: {
      id: api.nodes.api.id,
      name: "WebSocketApi",
      executionArn: api.nodes.api.executionArn,
    },
    lambda: { function: "src/lambdas/Authorizer/index.handler" },
    transform: {
      authorizer: {
        authorizerResultTtlInSeconds: 0,
      }
    }
  })

  api.route("$connect", "src/lambdas/websocket/index.connect", {
    transform: {
      route: {
        authorizerId: authorizer.id,
      },
    },
  })
  api.route("$disconnect", "src/lambdas/websocket/index.disconnect")
  api.route("$default", "src/lambdas/websocket/index.defaultHandler")
  api.route("sendMessage", "src/lambdas/websocket/index.sendMessage")

  return { api }
}
```

```
✕  Failed
   WebSocketApiAuthorizer sst:aws:ApiGatewayV2Authorizer → Web
SocketApiAuthorizerAuthorizer aws:apigatewayv2:Authorizer
     sdk-v2/provider2.go:520: sdk.helper_schema: creating API
Gateway v2 Authorizer (gapsheet-wj-WebSocketApiAuthorizerAutho
rizer): operation error ApiGatewayV2: CreateAuthorizer, https
response error StatusCode: 400, RequestID: 7c7ef069-7ad9-4f0b-
bae1-5364ec4e38c1, BadRequestException: AuthorizerResultTtlInS
econds cannot be set for WEBSOCKET protocol Apis.: provider=aw
s@6.59.1

   WebSocketApiAuthorizer sst:aws:ApiGatewayV2Authorizer → Web
SocketApiAuthorizerAuthorizer aws:apigatewayv2:Authorizer
   StatusCode: 400, RequestID: 7c7ef069-7ad9-4f0b-bae1-5364ec4
e38c1, BadRequestException: AuthorizerResultTtlInSeconds canno
t be set for WEBSOCKET protocol Apis.
```

Would love some help from someone who knows more about Pulumi and SST internals.

Edit: to be clear, `authorizerResultTtlInSeconds` seems to be required, meaning I'm not sure how to add an authorizer to a WEBSOCKET type APIGW.

**Comment by wilfredjonathanjames** [comment]
Closer. This creates the authorizer and attaches it to the $connect route. All connections are rejected. Debugging.

```typescript
/// <reference path="../.sst/platform/config.d.ts" />

import { Input } from "@pulumi/pulumi"
import { formatPulumiResourceName } from "./utils"

export default async function configureWebSockets(
  vpc: sst.aws.Vpc,
  db: sst.aws.Postgres,
  authorizerFunction: sst.aws.Function,
) {
  const api = new sst.aws.ApiGatewayWebSocket("WebSocket", {
    transform: {
      route: {
        handler: {
          link: [db],
          vpc,
        },
      },
    },
  })

  const authorizer = new aws.apigatewayv2.Authorizer(
    formatPulumiResourceName("WebSocketApiAuthorizer"),
    {
      apiId: api.nodes.api.id,
      authorizerType: "REQUEST",
      authorizerUri: authorizerFunction.nodes.function
        .invokeArn as Input<string>,
      identitySources: ["route.request.header.Auth"],
    },
  )

  api.route("$connect", "src/lambdas/websocket/index.connect", {
    transform: {
      route: {
        authorizerId: authorizer.id,
        authorizationType: "CUSTOM",
      },
    },
  })
  api.route("$disconnect", "src/lambdas/websocket/index.disconnect")
  api.route("$default", "src/lambdas/websocket/index.defaultHandler")
  api.route("sendMessage", "src/lambdas/websocket/index.sendMessage")

  return { api }
}
```

**Comment by wilfredjonathanjames** [comment]
Leaving this here. I'm getting the following CloudWatch error.

```
{
    "requestId": "BmPtwEVLIAMFTOg=",
    "ip": "[...]",
    "requestTime": "21/Nov/2024:12:51:10 +0000",
    "httpMethod": "-",
    "routeKey": "$connect",
    "status": "500",
    "protocol": "-",
    "errorMessage": "Internal server error",
    "path": "-",
    "authorizerPrincipalId": "-",
    "user": "-",
    "caller": "-",
    "validationErrorString": "-",
    "errorResponseType": "API_CONFIGURATION_ERROR",
    "integrationErrorMessage": "-",
    "responseLength": "-",
    "integrationStatus": "-",
    "integrationLatency": "-"
}
```

Putting a pause on this to try validation in the $connect lambda.

**Comment by jayair** [comment]
We might need this for our auth @fwang?

**Comment by wilfredjonathanjames** [comment]
While I'd prefer to use an authorizer, I've had success validating in the $connect handler. Considering the auth mechanism for WebSockets differs a little from other endpoints I'm happy enough to go with this for now. With that said I'd really rather split it out if or when we get support for this.

**Comment by culda** [comment]
Just running into this but I don't think it has anything to do with sst.
My $connect is a basic lambda handler
```
export const handler = async(event, context) => {
  return {"statusCode" : 200}
};
```

But socket won't upgrade conn even though the connect handler returns 200

**Comment by culda** [comment]
I created a Websocket API manually in the console, got it to be identical to the one deployed via SST, and I can only connect to the manual one. I think I'm going crazy over this, they are truly identical, bit by bit, except the resource ID. But the sst one gives 403 and the other one connects.

**Comment by culda** [comment]
The only discrepancy in the CW logs is this entry suggesting the lambda tries to make an external call, that's so strange.
```
025/02/24 21:52:45 ERROR run failed err="dial tcp :443: connect: connection refused"
```

**Comment by culda** [comment]
I think the link was causing the 443 conn refused attempt. It works now. Will look into this more tmrw.
```
    wsApi.route("$connect", {
      handler: "functions/connect.handler",
      // link: [wsConnTable],
    })

```

**Comment by jayair** [comment]
@culda did you find something?

**Comment by culda** [comment]
> @culda did you find something?

I didn't get around to the authorizer yet. Probably tomorrow. My other issues were irrelevant to this thread actually and are resolved now.
